File: src/leetcode/e121/BestTimeToBuyAndSellStock.go
package e121

// LeetCode 121
func maxProfit(prices []int) int {
	profit := 0
	minBuy := prices[0]
	for i := 1; i < len(prices); i++ {
		currProfit := prices[i] - minBuy
		profit = maxInt(currProfit, profit)
		minBuy = minInt(prices[i], minBuy)
	}
	return profit
}

func maxInt(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func minInt(a, b int) int {
	if a < b {
		return a
	}
	return b
}

File: src/leetcode/e121/BestTimeToBuyAndSellStock.java
/*
 * Copyright (c) 2024 THGEEK.COM. All rights reserved.
 */

package leetcode.e121;

/**
 * LeetCode 121
 *
 * @author taohong on 05/09/2018
 */
public class BestTimeToBuyAndSellStock {
    public int maxProfit(int prices[]) {
        int minPrice = Integer.MAX_VALUE;
        int maxProfit = 0;
        for (int i = 0; i < prices.length; i++) {
            if (prices[i] < minPrice) {
                minPrice = prices[i];
            } else if (prices[i] - minPrice > maxProfit) {
                maxProfit = prices[i] - minPrice;
            }
        }
        return maxProfit;
    }
}
